src/services/htmlCompletion.ts

```diff
@@ -9,407 +9,435 @@ import { HTMLDocument, Node } from '../parser/htmlParser';
-		}
+			function addNodeDataAttributes(node: Node) {
+				node.attributeNames.forEach(attr => {
+					if (startsWith(attr, dataAttr) && !dataAttributes[attr]) {
+						dataAttributes[attr] = attr + '="$1"';
+					}
+				});
+				node.children.forEach(child => addNodeDataAttributes(child));
+			}
 
-		if (htmlDocument) {
-			htmlDocument.roots.forEach(root => addNodeDataAttributes(root));
+			if (htmlDocument) {
+				htmlDocument.roots.forEach(root => addNodeDataAttributes(root));
+			}
+			Object.keys(dataAttributes).forEach(attr => result.items.push({
+				label: attr,
+				kind: CompletionItemKind.Value,
+				textEdit: TextEdit.replace(range, dataAttributes[attr]),
+				insertTextFormat: InsertTextFormat.Snippet
+			}));
 		}
-		Object.keys(dataAttributes).forEach(attr => result.items.push({
-			label: attr,
-			kind: CompletionItemKind.Value,
-			textEdit: TextEdit.replace(range, dataAttributes[attr]),
-			insertTextFormat: InsertTextFormat.Snippet
-		}));
-	}
 
-	function collectAttributeValueSuggestions(valueStart: number, valueEnd: number = offset): CompletionList {
-		let range: Range;
-		let addQuotes: boolean;
-		if (offset > valueStart && offset <= valueEnd && text[valueStart] === '"') {
-			// inside attribute
-			if (valueEnd > offset && text[valueEnd - 1] === '"') {
-				valueEnd--;
+		function collectAttributeValueSuggestions(valueStart: number, valueEnd: number = offset): CompletionList {
+			let range: Range;
+			let addQuotes: boolean;
+			if (offset > valueStart && offset <= valueEnd && text[valueStart] === '"') {
+				// inside attribute
+				if (valueEnd > offset && text[valueEnd - 1] === '"') {
+					valueEnd--;
+				}
+				let wsBefore = getWordStart(text, offset, valueStart + 1);
+				let wsAfter = getWordEnd(text, offset, valueEnd);
+				range = getReplaceRange(wsBefore, wsAfter);
+				addQuotes = false;
+			} else {
+				range = getReplaceRange(valueStart, valueEnd);
+				addQuotes = true;
 			}
-			let wsBefore = getWordStart(text, offset, valueStart + 1);
-			let wsAfter = getWordEnd(text, offset, valueEnd);
-			range = getReplaceRange(wsBefore, wsAfter);
-			addQuotes = false;
-		} else {
-			range = getReplaceRange(valueStart, valueEnd);
-			addQuotes = true;
-		}
-		let tag = currentTag.toLowerCase();
-		let attribute = currentAttributeName.toLowerCase();
-		tagProviders.forEach(provider => {
-			provider.collectValues(tag, attribute, value => {
-				let insertText = addQuotes ? '"' + value + '"' : value;
-				result.items.push({
-					label: value,
-					filterText: insertText,
-					kind: CompletionItemKind.Unit,
-					textEdit: TextEdit.replace(range, insertText),
-					insertTextFormat: InsertTextFormat.PlainText
+
+			let tag = currentTag.toLowerCase();
+			let attribute = currentAttributeName.toLowerCase();
+			let value = extractAttributeValue(scanner.getTokenText());
+
+			completionParticipants.forEach(participant => {
+				if (participant.onHtmlAttributeValue) {
```

Comment:
Move outside of forEach

Reply:
Why -- shouldn't each of them be checked?
Let's say you have 3 participants and only two of them have `onHtmlAttributeValue`. If I do not do the check inside the foreach (or for let) `participant.onHtmlAttributeValue` can be undefined.

Reply:
oh, sorry, you are right of course.